openarena: cleanup

drupol opened the pull request:
In this PR, the following changes have been made:

- Add missing hook phases
- Add desktop item with a nice icon
- Use `finalAttrs` pattern
- Sort attributes and parameters naturally
- Use `fetchzip` (:warning: _the source URL has changed!, basically the previous URL was a redirection to this one_)
- Support Darwin arch
- Use `autoPatchelfHook`
- Reduce the size of the closure by copying only what's needed
- Update the licence
- Add `meta.mainProgram` attribute
- Clean-up the nixos module, remove the hardcoded version, add `package` config option

CC: @alicerum

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

drupol committed:
openarena: cleanup

- Add missing hook phases
- Add desktop item
- Use `finalAttrs` pattern
- Sort attributes and parameters naturally
- Use `fetchzip`
- Reduce the size of the closure by copying only the necessary files
- Use `autoPatchelfHook`
- Support Darwin arch
- Update license
- Add `meta.mainProgram` attribute

tobim commented on pkgs/games/openarena/default.nix:
```diff
@@ -1,43 +1,95 @@
     homepage = "http://openarena.ws/";
-    maintainers = [ lib.maintainers.wyvie ];
-    platforms = [ "i386-linux" "x86_64-linux" ];
     license = lib.licenses.gpl2;
```
Can you update the license to `gpl2Plus` while you are at it?
https://github.com/OpenArena/engine/blob/master/ui/menudef.h#L9-L10

drupol replied:
Sure, will do.

drupol replied:
Done.

drupol committed:
nixos/openarena: cleanup

- Remove `with lib;`
- Add `package` config option
- Adjust paths accordingly in `ExecStart`

reckenrode commented on pkgs/games/openarena/default.nix:
```diff
@@ -1,43 +1,96 @@
+    license = lib.licenses.gpl2Plus;
+    mainProgram = "openarena";
+    maintainers = with lib.maintainers; [ wyvie ];
+    platforms = [ "i386-linux" "x86_64-linux" ] ++ lib.platforms.darwin;
```
I’m not sure this will even actually work, but the upstream app bundles are for x86 and PPC (no 64-bit builds as far as I can tell).

```suggestion
    platforms = [ "i386-linux" "i686-darwin" "x86_64-linux" ];
```

TheBrainScrambler commented on pkgs/games/openarena/default.nix:
```diff
@@ -1,43 +1,96 @@
+, autoPatchelfHook
+, curl
+, fetchzip
+, lib
```
https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md the very last line:
"Arguments should be listed in the order they are used, with the exception of lib, which always goes first."

Shouldn't lib be the first argument ?

drupol replied:
Done.

drupol committed:
openarena: build from source

reckenrode commented:
This patch allows it to build and run, but it’s pretty broken. It just keeps resetting back to the main menu.

The platforms should probably also be `x86_64-darwin` and `x86_64-linux` (and maybe `i686-linux`). As far as I can tell, there’s no upstream support for aarch64 at all. I left it as-is for now.

I also moved the libraries from `nativeBuildInputs` to `buildInputs`.

```patch
diff --git a/pkgs/games/openarena/default.nix b/pkgs/games/openarena/default.nix
index b820a5f21869..079f848fe1a1 100644
--- a/pkgs/games/openarena/default.nix
+++ b/pkgs/games/openarena/default.nix
@@ -3,17 +3,19 @@
 , fetchFromGitHub
 , fetchzip
 , fetchpatch
-, curl
 , freetype
 , libglvnd
 , libogg
 , libvorbis
 , libxmp
-, makeBinaryWrapper
 , openal
-, pkg-config
 , SDL2
 , speex
+, Carbon
+, Cocoa
+, curl
+, pkg-config
+, makeBinaryWrapper
 , which
 , copyDesktopItems
 , autoPatchelfHook
@@ -55,18 +57,24 @@ stdenv.mkDerivation (finalAttrs: {
     })
   ];
 
-  nativeBuildInputs = [
-    curl
+  buildInputs = [
     freetype
     libglvnd
     libogg
     libvorbis
     libxmp
-    makeBinaryWrapper
     openal
-    pkg-config
     SDL2
     speex
+  ] ++ lib.optionals stdenv.isDarwin [
+    Carbon
+    Cocoa
+  ];
+
+  nativeBuildInputs = [
+    curl
+    makeBinaryWrapper
+    pkg-config
     which
   ] ++ lib.optionals stdenv.isLinux [
     copyDesktopItems
@@ -80,8 +88,17 @@ stdenv.mkDerivation (finalAttrs: {
     "USE_FREETYPE=1"
     "USE_OPENAL_DLOPEN=0"
     "USE_CURL_DLOPEN=0"
+    "ARCH=${stdenv.hostPlatform.linuxArch}"
   ];
 
+  preBuild = lib.optionalString stdenv.isDarwin ''
+    # Use openal and SDL2 from nixpkgs
+    substituteInPlace Makefile \
+      --replace '-framework OpenAL' -lopenal \
+      --replace 'LIBSDLMAIN=$(B)/libSDL2main.a' "" \
+      --replace '$(LIBSDIR)/macosx/libSDL2-2.0.0.dylib' '$(SDL_LIBS)'
+  '';
+
   installTargets = [ "copyfiles" ];
   installFlags = [ "COPYDIR=$(out)/share/openarena" ];
 
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index c77c662369e9..671779f9b9d5 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -37896,7 +37896,9 @@ with pkgs;
 
   onscripter-en = callPackage ../games/onscripter-en { };
 
-  openarena = callPackage ../games/openarena { };
+  openarena = callPackage ../games/openarena {
+    inherit (darwin.apple_sdk.frameworks) Carbon Cocoa;
+  };
 
   opendungeons = callPackage ../games/opendungeons { };
 
```

miangraham commented:
Result of `nixpkgs-review pr 249902` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>openarena</li>
  </ul>
</details>

drupol commented:
Thanks @reckenrode.

As discussed on Matrix, the effort to get it running on Darwin is way too big at the moment.
Therefore, I will not add the patch you made, it will only be for Linux, hoping that at some point, the upstream will support Darwin properly.

I'm going to close this PR and continue the work in a new PR at #250105.
